### PR TITLE
Use same `FxHashMap` in `rustdoc-json-types` and `librustdoc`.

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
The original motivation was me trying to remove the `#![allow(rustc::default_hash_types)]`, as after #108626, we should be using `FxHashMap` here. I then realized I should also be able to remove the `.into_iter().collect()`, as we no longer need to convert from `FxHashMap` to `std::HashMap`.

However, this didn't work, and I got the following error

```
error[E0308]: mismatched types
   --> src/librustdoc/json/mod.rs:235:13
    |
235 |             index,
    |             ^^^^^ expected `rustc_hash::FxHasher`, found `FxHasher`
    |
    = note: `FxHasher` and `rustc_hash::FxHasher` have similar names, but are actually distinct types
note: `FxHasher` is defined in crate `rustc_hash`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/rustc-hash-1.1.0/src/lib.rs:60:1
note: `rustc_hash::FxHasher` is defined in crate `rustc_hash`
   --> /home/alona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustc-hash-1.1.0/src/lib.rs:60:1
    |
60  | pub struct FxHasher {
    | ^^^^^^^^^^^^^^^^^^^
    = note: perhaps two different versions of crate `rustc_hash` are being used?
```

This is because `librustdoc` got it's `FxHashMap` via the sysroot `rustc-data-strucures`, whereas `rustdoc-json-types` got it via `rustc-hash` on crates.io.

To avoid this, `rustdoc-json-types` now uses
`#![feature(rustc_private)]` to load the same version as `librustdoc`.

However, this needs to be placed behind a feature, as `rustdoc-json-types` is also dependency of `src/tools/jsondocck`, which means need needs to be buildable without nightly features.

r? @jyn514

### CODE REVIEW

**Comment by jyn514** [comment]
> However, this needs to be placed behind a feature, as rustdoc-json-types is also dependency of src/tools/jsondocck, which means need needs to be buildable without nightly features.

Hmm, I thought the motivation was that you want to be able to Auto publish rustdoc-json-types to crates.io so people can use it? Is jsondocck built with the beta compiler or something like that?

**Review by jyn514 — APPROVED** [review]
👍 this looks good, thanks! r=me when CI passes

**Comment by jyn514** [comment]
> `Building tool jsondoclint (stage0)`

This bit means it's built with beta :)

@bors r+ rollup

**Comment by Nilstrieb** [comment]
https://github.com/rust-lang/rust/pull/110073#issuecomment-1500815909

@bors r-

**Comment by aDotInTheVoid** [comment]
Running `./x dist` with `profile = "compiler"`, I can't reproduce the issue locally after rebasing.

@bors try

**Comment by jyn514** [comment]
```
/bin/sh: line 1: 56093 Segmentation fault: 11  ../../../bin/llvm-tblgen -gen-pseudo-lowering -I /Users/runner/work/rust/rust/src/llvm-project/llvm/lib/Target/AArch64 -I/Users/runner/work/rust/rust/build/x86_64-apple-darwin/llvm/build/include -I/Users/runner/work/rust/rust/src/llvm-project/llvm/include -I /Users/runner/work/rust/rust/src/llvm-project/llvm/lib/Target -no-warn-on-unused-template-args /Users/runner/work/rust/rust/src/llvm-project/llvm/lib/Target/AArch64/AArch64.td --write-if-changed -o /Users/runner/work/rust/rust/build/x86_64-apple-darwin/llvm/build/lib/Target/AArch64/AArch64GenMCPseudoLowering.inc
```
Well that's concerning.

@bors retry

**Comment by matthiaskrgr** [comment]
@bors r- 
failed in https://github.com/rust-lang/rust/pull/110360

**Comment by aDotInTheVoid** [comment]
@jyn514 any idea why `dist-x86_64-linux` would pass (in try), but `dist-arm-linux` and  `dist-armhf-linux` would fail in rollup.

```
error[E0464]: multiple candidates for `rmeta` dependency `rustc_hash` found
 --> src/rustdoc-json-types/lib.rs:8:1
  |
8 | extern crate rustc_hash;
  | ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: candidate #1: /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/arm-unknown-linux-gnueabi/lib/librustc_hash-0e35be53e73dd6f8.rmeta
  = note: candidate #2: /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/arm-unknown-linux-gnueabi/lib/librustc_hash-4e38a0a2586a3427.rmeta
```

**Comment by jyn514** [comment]
I see `rustc-hash` show up two times in the build logs:
1. https://github.com/rust-lang-ci/rust/actions/runs/4708387061/jobs/8350792846#step:26:6797 while building the stage1 compiler: https://github.com/rust-lang-ci/rust/actions/runs/4708387061/jobs/8350792846#step:26:6632
2. https://github.com/rust-lang-ci/rust/actions/runs/4708387061/jobs/8350792846#step:26:58752 while building the stage 2 compiler: https://github.com/rust-lang-ci/rust/actions/runs/4708387061/jobs/8350792846#step:26:58720

Those both seem correct to me, I'm not sure what could be going wrong :/

**Comment by JohnCSimon** [comment]
triage: 
@aDotInTheVoid - seems like this was waiting on author when it was meant for review, correct

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

@rustbot ready

**Comment by jyn514** [comment]
@JohnCSimon no, it's not ready for review. it fails only in a full bors merge, which runs more tests than PR CI or `bors try`

**Comment by aDotInTheVoid** [comment]
Yep, I need to figure out whats going on with this. I've seen the same error transitently locally, but only when my `build/` was full from a load of other random dev work. Runing `x clean` fixed this though. One day I'll pick this up again when theirs nothing more important to do (or more honestly knowing me, when my brain decides for no apparent reason that i *realy must know* what's going on with that weird linker error(

I might see if approving works if the queue happens to be empty on Christmas (and no-one else it trying the same thing lol).

Sorry for the confusion

@rustbot author

**Comment by Nilstrieb** [comment]
uh, this is marked as pending.

**Pull request closed without merging**
